google/cloud/internal/async_retry_loop.h

```diff
@@ -59,6 +59,109 @@ struct FutureValueType<future<T>> {
  * implementation of these stubs is very easy too.
  *
  * This class implements the retry loop for such an RPC.
+ *
+ * @par Cancellation and Thread Safety
+ *
+ * This class supports cancelling a retry loop (best-effort as most cancels
+ * with gRPC). Without cancels this class would require no synchronization, as
+ * each request, backoff timer, and their callbacks can only occur in sequence.
+ * Cancel requests, however, may be invoked by any thread and require some form
+ * of synchronization. The basic idea is that the class maintains a future
+ * (called `pending_operation_`) that represents the current pending operation,
+ * i.e., the current gRPC request or backoff timer. Cancelling the loop may
+ * require cancelling the pending operation.  Using such a `pending_operation_`
+ * future is prone to subtle race conditions. This section explains how we
+ * ensure safety.
+ *
+ * @par The Race Condition
+ *
+ * The most common approach to solve race conditions is to use some kind of
+ * lock, unfortunately this does not work in this case. Consider this code:
+ *
+ * @code
+ * 1: void AsyncRetryLoopImpl::F() {
+ * 2:  auto self = shared_from_this();
+ * 3:  future<...> f = create_future();
+ * 4:  std::unique_lock lk(mu_);
+ * 5:  if (cancelled_) return HandleCancel();
+ * 6:  auto pending = f.then([self](auto g) { self->Callback(); }
+ * 7:  this->pending_operation_ = pending;
+ * 8: }
+ * 9:
+ * 10: void AsyncRetryLoopImpl::Cancel() {
+ * ....
+ * 11:   this->pending_operation_.cancel();
+ * ....
+ * }
+ * 12:
+ * 13: void AsyncRetryLoopImpl::Callback() {
+ * ....
+ * 14:   F();
+ * ...
+ * }
+ * @endcode
+ *
+ * Because futures are immediately active, setting up the callback on line (6)
+ * can result in an immediate call to `Callback()`.  Since the mutex is held in
+ * line (4) that can result in a deadlock as the callback may invoke `F()`
+ * in line (14) which would want to acquire the mutex.
+ *
+ * One could solve this deadlock using a recursive mutex, but that still leaves
+ * a second problem: as the stack unwinds the `pending_operation_` member
+ * variable is set, on line (7), to the **first** operation, and we want it to
+ * remain set to the **last** operation.
+ *
+ * @par Solution
+ *
+ * We keep a counter reflecting the current iteration of the retry loop. This
+ * counter is incremented before starting a request and before starting a
+ * backoff timer.
```

Comment:
What do you think about changing all "iteration"s to "operation"? (e.g. `StartIteration()` -> `StartOperation()`, etc.)

"iteration" does makes sense when we think about the retry loop as `[request, backoff, request, backoff, request, ...]`, but I think "operation" is even clearer.

If you don't like that suggestion, maybe we could explicitly define the loop. Like: "The retry loop is an alternating sequence of requests and backoffs. We keep a counter reflecting the current iteration....". That way we won't think of one iteration as a request + a backoff.

Reply:
Good idea, fixed.